File: app/src/main/java/ch/admin/bag/dp3t/checkin/checkinflow/CheckOutFragment.java
package ch.admin.bag.dp3t.checkin.checkinflow;

import android.app.TimePickerDialog;
import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.lifecycle.ViewModelProvider;

import java.util.Calendar;

import org.crowdnotifier.android.sdk.CrowdNotifier;
import org.crowdnotifier.android.sdk.model.VenueInfo;

import ch.admin.bag.dp3t.R;
import ch.admin.bag.dp3t.checkin.CrowdNotifierViewModel;
import ch.admin.bag.dp3t.checkin.models.CheckInState;
import ch.admin.bag.dp3t.checkin.models.DiaryEntry;
import ch.admin.bag.dp3t.checkin.storage.DiaryStorage;
import ch.admin.bag.dp3t.checkin.utils.CrowdNotifierReminderHelper;
import ch.admin.bag.dp3t.checkin.utils.NotificationHelper;
import ch.admin.bag.dp3t.extensions.VenueInfoExtensionsKt;
import ch.admin.bag.dp3t.util.StringUtil;

public class CheckOutFragment extends Fragment {

	public static final String TAG = CheckOutFragment.class.getCanonicalName();

	private static final long ONE_DAY_IN_MILLIS = 24 * 60 * 60 * 1000L;

	private CrowdNotifierViewModel viewModel;
	private VenueInfo venueInfo;
	private CheckInState checkInState;

	private TextView titleTextView;
	private TextView subtitleTextView;
	private View doneButton;
	private View cancelButton;
	private TextView fromTime;
	private TextView toTime;
	private TextView dateTextView;
	private Button hideInDiaryButton;

	public CheckOutFragment() { super(R.layout.fragment_check_out_and_edit); }

	public static CheckOutFragment newInstance() {
		return new CheckOutFragment();
	}

	@Override
	public void onCreate(@Nullable Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		viewModel = new ViewModelProvider(requireActivity()).get(CrowdNotifierViewModel.class);
		checkInState = viewModel.getCheckInState();
		if (checkInState != null) {
			venueInfo = checkInState.getVenueInfo();
		}
		checkIfAutoCheckoutHappened();
	}

	@Override
	public void onStart() {
		super.onStart();
		checkIfAutoCheckoutHappened();
	}

	private void checkIfAutoCheckoutHappened() {
		if (viewModel.getCheckInState() == null) {
			popBackToHomeFragment();
		}
	}

	@Override
	public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
		titleTextView = view.findViewById(R.id.check_out_fragment_title);
		subtitleTextView = view.findViewById(R.id.check_out_fragment_subtitle);
		doneButton = view.findViewById(R.id.check_out_fragment_done_button);
		cancelButton = view.findViewById(R.id.check_out_fragment_cancel_button);
		fromTime = view.findViewById(R.id.check_out_fragment_from_text_view);
		toTime = view.findViewById(R.id.check_out_fragment_to_text_view);
		dateTextView = view.findViewById(R.id.check_out_fragment_date);
		hideInDiaryButton = view.findViewById(R.id.edit_diary_entry_hide_from_diary_button);

		titleTextView.setText(venueInfo.getTitle());
		subtitleTextView.setText(VenueInfoExtensionsKt.getSubtitle(venueInfo));

		checkInState.setCheckOutTime(System.currentTimeMillis());
		refreshTimeTextViews();

		fromTime.setOnClickListener(v -> showTimePicker(true));
		toTime.setOnClickListener(v -> showTimePicker(false));

		hideInDiaryButton.setVisibility(View.GONE);

		doneButton.setOnClickListener(v -> {
			CrowdNotifierReminderHelper.removeAllReminders(getContext());
			saveEntry();
			NotificationHelper notificationHelper = NotificationHelper.getInstance(getContext());
			notificationHelper.stopOngoingNotification();
			notificationHelper.removeReminderNotification();
			popBackToHomeFragment();
		});
		cancelButton.setOnClickListener(v -> requireActivity().getSupportFragmentManager().popBackStack());
	}

	private void refreshTimeTextViews() {
		fromTime.setText(StringUtil.getHourMinuteTimeString(checkInState.getCheckInTime(), "  :  "));
		toTime.setText(StringUtil.getHourMinuteTimeString(checkInState.getCheckOutTime(), "  :  "));
		dateTextView.setText(StringUtil.getCheckOutDateString(getContext(), checkInState.getCheckInTime(),
				checkInState.getCheckOutTime()));
	}

	private void showTimePicker(boolean isFromTime) {
		Calendar time = Calendar.getInstance();
		if (isFromTime) {
			time.setTimeInMillis(checkInState.getCheckInTime());
		} else {
			time.setTimeInMillis(checkInState.getCheckOutTime());
		}
		int hour = time.get(Calendar.HOUR_OF_DAY);
		int minute = time.get(Calendar.MINUTE);
		TimePickerDialog timePicker;
		timePicker = new TimePickerDialog(getContext(), (picker, selectedHour, selectedMinute) -> {
			time.set(Calendar.HOUR_OF_DAY, selectedHour);
			time.set(Calendar.MINUTE, selectedMinute);
			if (isFromTime) {
				checkInState.setCheckInTime(time.getTimeInMillis());
			} else {
				checkInState.setCheckOutTime(time.getTimeInMillis());
			}
			if (checkInState.getCheckOutTime() < checkInState.getCheckInTime()) {
				checkInState.setCheckOutTime(checkInState.getCheckOutTime() + ONE_DAY_IN_MILLIS);
			} else if (checkInState.getCheckInTime() + ONE_DAY_IN_MILLIS < checkInState.getCheckOutTime()) {
				checkInState.setCheckOutTime(checkInState.getCheckOutTime() - ONE_DAY_IN_MILLIS);
			}
			refreshTimeTextViews();
		}, hour, minute, true);

		timePicker.show();
	}

	private void saveEntry() {
		long checkIn = checkInState.getCheckInTime();
		long checkOut = checkInState.getCheckOutTime();
		long id = CrowdNotifier.addCheckIn(checkIn, checkOut, venueInfo, getContext());
		DiaryStorage.getInstance(getContext()).addEntry(new DiaryEntry(id, checkIn, checkOut, venueInfo));
		viewModel.setCheckInState(null);
	}

	private void popBackToHomeFragment() {
		requireActivity().getSupportFragmentManager().popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE);
	}

}

File: app/src/main/java/ch/admin/bag/dp3t/checkin/checkinflow/CheckOutFragment.kt
package ch.admin.bag.dp3t.checkin.checkinflow

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.activityViewModels
import ch.admin.bag.dp3t.R
import ch.admin.bag.dp3t.checkin.CrowdNotifierViewModel
import ch.admin.bag.dp3t.checkin.diary.CheckinTimeHelper
import ch.admin.bag.dp3t.checkin.models.CheckInState
import ch.admin.bag.dp3t.checkin.models.DiaryEntry
import ch.admin.bag.dp3t.checkin.storage.DiaryStorage
import ch.admin.bag.dp3t.checkin.utils.CrowdNotifierReminderHelper
import ch.admin.bag.dp3t.checkin.utils.NotificationHelper
import ch.admin.bag.dp3t.databinding.FragmentCheckOutAndEditBinding
import ch.admin.bag.dp3t.extensions.getSubtitle
import ch.admin.bag.dp3t.extensions.getSwissCovidLocationData
import org.crowdnotifier.android.sdk.CrowdNotifier
import org.crowdnotifier.android.sdk.model.VenueInfo

class CheckOutFragment : Fragment() {

	companion object {
		private val TAG = CheckOutFragment::class.java.canonicalName

		@JvmStatic
		fun newInstance() = CheckOutFragment()
	}

	private val viewModel: CrowdNotifierViewModel by activityViewModels()

	private lateinit var venueInfo: VenueInfo
	private lateinit var checkInState: CheckInState

	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		checkIfAutoCheckoutHappened()
		checkInState = viewModel.checkInState?.copy(checkOutTime = System.currentTimeMillis()) ?: return
		venueInfo = checkInState.venueInfo
	}

	override fun onStart() {
		super.onStart()
		checkIfAutoCheckoutHappened()
	}

	private fun checkIfAutoCheckoutHappened() {
		if (viewModel.checkInState == null) {
			popBackToHomeFragment()
		}
	}

	override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View {
		return FragmentCheckOutAndEditBinding.inflate(inflater, container, false).apply {
			checkoutTitle.text = venueInfo.title
			checkoutSubtitle.setText(venueInfo.getSubtitle())

			toolbarDoneButton.setOnClickListener { performCheckout() }
			toolbarCancelButton.setOnClickListener { requireActivity().supportFragmentManager.popBackStack() }
			checkoutPrimaryButton.setText(R.string.checkout_button_title)
			checkoutPrimaryButton.setOnClickListener { performCheckout() }

			checkoutTimeArrival.setDateTime(checkInState.checkInTime)
			checkoutTimeArrival.setOnDateTimeChangedListener {
				checkInState.checkInTime = checkoutTimeArrival.getSelectedUnixTimestamp()
			}

			checkoutTimeDeparture.setDateTime(checkInState.checkOutTime)
			checkoutTimeDeparture.setOnDateTimeChangedListener {
				checkInState.checkOutTime = checkoutTimeDeparture.getSelectedUnixTimestamp()
			}
		}.root
	}

	private fun performCheckout() {
		checkInState.run {
			if (checkInTime > checkOutTime) {
				// swap arrival and departure time
				checkInTime = checkOutTime.also { checkOutTime = checkInTime }
			}
		}

		val hasOverlapWithOtherCheckin =
			CheckinTimeHelper.checkForOverlap(checkInState.checkInTime, checkInState.checkOutTime, requireContext())
		if (hasOverlapWithOtherCheckin) {
			CheckinTimeHelper.showSavingNotPossibleDialog(
				getString(R.string.checkout_overlapping_alert_description),
				requireContext()
			)
			return
		}

		val checkinDuration = checkInState.checkOutTime - checkInState.checkInTime
		val maxCheckinTime = checkInState.venueInfo.getSwissCovidLocationData().automaticCheckoutDelaylMs
		if (checkinDuration > maxCheckinTime) {
			val dialogText = CheckinTimeHelper.getMaxCheckinTimeExceededMessage(maxCheckinTime, requireContext())
			CheckinTimeHelper.showSavingNotPossibleDialog(dialogText, requireContext())
			return
		}

		CrowdNotifierReminderHelper.removeAllReminders(context)
		saveEntry()
		val notificationHelper = NotificationHelper.getInstance(context)
		notificationHelper.stopOngoingNotification()
		notificationHelper.removeReminderNotification()
		popBackToHomeFragment()
	}

	private fun saveEntry() {
		val checkIn = checkInState.checkInTime
		val checkOut = checkInState.checkOutTime
		val id = CrowdNotifier.addCheckIn(checkIn, checkOut, venueInfo, context)
		DiaryStorage.getInstance(context).addEntry(DiaryEntry(id, checkIn, checkOut, venueInfo))
		viewModel.checkInState = null
	}

	private fun popBackToHomeFragment() {
		requireActivity().supportFragmentManager.popBackStack(null, FragmentManager.POP_BACK_STACK_INCLUSIVE)
	}

}